### DISCUSSION TITLE
Plot netcdf data with specific shapefile

### DISCUSSION THREAD

**Discussion by Lij-16** [jddingemanse/awtiCodeDev#21] [Q&A] [answered]
I have 12 CMIP6 data in NetCDF format. I want to spatial plot these models with my study area(shapefile) in one page. How can I plot? Do you have python scripts/tools for this purpose? Please, help me anyone.
Thank you.

**Comment by jddingemanse** [comment, marked as answer]
Nice question! First of all, if you want to work with both NetCDF files and shapefiles, you should have the packages that can load those files, for example xarray and netcdf4 to work with .nc files, and geopandas to work with .shp files (or .zip files with .shp files in them).

If that is the case, then the following pieces of code might help you.

### Loading data and pre-processing netcdf data
First, import the necessary packages, and load the data.
```
import xarray as xr
import geopandas as gpd
ds1 = xr.open_dataset('your_netcdf_file.nc') # This loads one of the netcdf files as xarray DataSet - you can do the same to load all 12 of them.
area = gpd.read_file('your_shape_file.shp') # Or shapefile.zip. This will load your shapefile as GeoDataFrame.
```
To plot data from the .nc file on a map, you need to make sure that you have your data with only two dimensions (longitude and latitude). You can check what the name of your data variable is with `ds1.data_var`, and then you can check the dimensions of this data variable with `ds1.[name_of_data_var].dims`. If your dimensions are for example (time,lat,lon), you either need to take an average over time, or only select one time.
To select an average over time:
```
timeAvg = ds1.[name_of_data_var].mean(dim='time') # Assuming the name of the dimension is 'time', and assuming that the only other dimensions are latitude and longitude.
```
This should give you, for one of your .nc files, an average over time for every latitude and longitude. It is quite easy to plot that:
```
timeAvg.plot()
```
The above simply plotted data for all latitudes and longitudes in the netcdf file. If you want to select only part of that, you should _slice_ your data first. For the sake of this example code I assume that timeAvg has only dimensions latitude and longitude, and that their names are `lat` and `lon`. Also, lets assume you want to slice for latitudes 6-8 and longitudes 14-18.
```
latmin,latmax,lonmin,lonmax = 6,8,14,18 # Define 4 variables, to use in slicing.
timeAvg[(timeAvg.lat>=latmin) & (timeAvg.lat<=latmax) , (timeAvg.lon>=lonmin) & (timeAvg.lon<=lonmax)]
```
### Plotting it together with a shapefile
If you want to plot the shapefile together with the netcdf data, you first need to make sure they are in the same Coordinate Reference System (CRS). Netcdf files usually are in degrees. The code for this CRS = `EPSG:4326`. You can check whether your shapefile is in this crs with `area.crs`. If the crs is wrong, you can change it with `area.to_crs('EPSG:4326')`.
Then, you can plot them together by plotting them on the same `matplotlib Axes`:
```
import matplotlib.pyplot as plt
fig,ax=plt.subplots() # Create an empty figure and axes
timeAvg.plot(ax=ax) # Plot the preprocessed netcdf data on that create ax
area.plot(ax=ax, facecolor='none', edgecolor = 'k') # Plot the shapefile on the same ax, with black lines and no fill color.
```
You talk about 12 netcdf files. If you want to plot them all together, you probably want to use 12 axes in 1 figure. After properly preprocessing your data (having 12 different timeAvg DataArrays), you can plot them on 12 different axes with something like this:
```
fig,((ax1,ax2,ax3),(ax4,ax5,ax6),(ax7,ax8,ax9),(ax10,ax11,ax12)) = plt.subplots(4,3) # Create empty figure with 12 Axes; 4 rows by 3 columns
timeAvg1.plot(ax=ax1)
# etcetera
```

Let me know where you get stuck!

**Comment by Lij-16** [comment]
Thank you very much for your quick response and clear explanation.I will do
it. Its very important for me.
